File: src/java/algorithms/sorting/QuickSort.java
package src.java.algorithms.sorting;
import src.java.data_structures.*;
import src.java.common.Tests;
import java.util.*;


public class QuickSort implements SortingAlgorithm
{

	public <T extends Comparable<T>> void execute(DataStructure<T> data, final Comparator<T> cmp)
	{
		if (data.size() > 1)
			sort(0, data.size(), data, cmp);
	}


	private <T extends Comparable<T>> void sort(final int begin, final int end, DataStructure<T> data, final Comparator<T> cmp)
	{
		if ((end - begin) > 1) {
			final int p = partitionate(begin, end, data, cmp);
			sort(begin, p, data, cmp);
			sort(p + 1, end, data, cmp);
		}
	}


	private <T extends Comparable<T>> int partitionate(int l, int r, DataStructure<T> data, final Comparator<T> cmp)
	{
		final int p = l;
		++l;
		--r;

	
		for (;;) {
			while (l < r && cmp.compare(data.get(l), data.get(p)) < 0)
				++l;
			while (r >= l && cmp.compare(data.get(p), data.get(r)) < 0)
				--r;

			if (l >= r)
				break;

			final T tmp = data.get(r);
			data.set(r, data.get(l));
			data.set(l, tmp);

			++l;
			--r;
		}


		final T tmp = data.get(r);
		data.set(r, data.get(p));
		data.set(p, tmp);
		return r;
	}


	public static void main(String[] args)
	{
		Tests.sort(args, new QuickSort());
	}
	
}

File: src/csharp/Algorithms/Sorting/QuickSort.cs
using Common;
using System;
using System.Collections.Generic;
using DataStructures;


namespace Algorithms.Sorting
{

	public class QuickSort : ISortingAlgorithm
	{

		public void Execute<T>(IDataStructure<T> data, Func<T, T, bool> cmp)
		{
			if (data.Count > 1)
				QSort(0, data.Count, data, cmp);
		}


		private void QSort<T>(int beg, int end, IDataStructure<T> data, Func<T, T, bool> cmp)
		{
			if ((end - beg) > 1) {
				int r = Partitionate(beg, end, data, cmp);
				QSort(r + 1, end, data, cmp);
				QSort(beg, r, data, cmp);
			}
		}


		private int Partitionate<T>(int beg, int end, IDataStructure<T> data, Func<T, T, bool> cmp)
		{
			int l = beg + 1;
			int r = end - 1;

			for (;;) {
				while (l < r && cmp(data[l], data[beg]))
					l += 1;
				while (r >= l && cmp(data[beg], data[r]))
					r -= 1;

				if (l >= r)
					break;

				T tmp = data[r];
				data[r] = data[l];
				data[l] = tmp;

				l += 1;
				r -= 1;
			}


			T aux = data[r];
			data[r] = data[beg];
			data[beg] = aux;
			return r;
		}


		public static int Main(string[] args)
		{
			return Tests.Sort(args, new QuickSort());
		}

	}
}
